File: day25/day25.cpp
#include <fstream>
#include <vector>
#include <fmt/format.h>

using namespace std;

struct Operand
{
	enum { REG, LIT } type;
	int64_t value;

	int64_t get(int64_t regs[]) const noexcept
	{
		return type == REG ? regs[value] : value;
	}
};

istream& operator>>(istream& in, Operand& o)
{
	string s;
	if (in >> s)
	{
		if ("abcd"s.find(s[0]) != string::npos)
		{
			o.type = Operand::REG;
			o.value = s[0]-'a';
		}
		else
		{
			o.type = Operand::LIT;
			o.value = stoi(s);
		}
	}
	return in;
}

struct Instruction
{
	enum { CPY, INC, DEC, JNZ, OUT } opcode;
	Operand op1;
	Operand op2;
};

istream& operator>>(istream& in, Instruction& i)
{
	string s;
	in >> s;
	if (s == "cpy")
	{
		i.opcode = Instruction::CPY;
		in >> i.op1;
		in >> i.op2;
	}
	else if (s == "inc")
	{
		i.opcode = Instruction::INC;
		in >> i.op1;
	}
	else if (s == "dec")
	{
		i.opcode = Instruction::DEC;
		in >> i.op1;
	}
	else if (s == "jnz")
	{
		i.opcode = Instruction::JNZ;
		in >> i.op1;
		in >> i.op2;
	}
	else if (s == "out")
	{
		i.opcode = Instruction::OUT;
		in >> i.op1;
	}
	else
	{
		in.setstate(ios::failbit);
	}
	return in;
}

struct Computer
{
	int64_t registers[4];

	bool execute(const vector<Instruction>& program, int64_t init) noexcept
	{
		registers[0] = init;
		fill(registers+1, registers+4, 0);
		vector<Computer> stack;
		int64_t last = -1;
		int64_t ip = 0;
		while (0 <= ip && (size_t)ip < program.size())
		{
			auto& instr = program[ip];
			switch (instr.opcode)
			{
			case Instruction::CPY:
				if (instr.op2.type == Operand::REG)
				{
					registers[instr.op2.value] = instr.op1.get(registers);
				}
				ip++;
				break;

			case Instruction::INC:
				if (instr.op1.type == Operand::REG)
				{
					registers[instr.op1.value]++;
				}
				ip++;
				break;

			case Instruction::DEC:
				if (instr.op1.type == Operand::REG)
				{
					registers[instr.op1.value]--;
				}
				ip++;
				break;

			case Instruction::JNZ:
				if (instr.op1.get(registers))
				{
					ip += instr.op2.get(registers);
				}
				else
				{
					ip++;
				}
				break;

			case Instruction::OUT:
			{
				int64_t value = instr.op1.get(registers);
				if (value == last)
				{
					return false;
				}
				last = value;

				while (!stack.empty())
				{
					auto& back = stack.back();
					if (back == *this)
					{
						return true;
					}

					if (back > *this)
					{
						stack.pop_back();
					}
					else
					{
						break;
					}
				}
				stack.push_back(*this);
				ip++;
				break;
			}

			default:
				abort();
			}
		}
		return false;
	}

	bool operator==(const Computer& c) const noexcept
	{
		for (int i = 0; i < 4; i++)
		{
			if (registers[i] != c.registers[i])
				return false;
		}
		return true;
	}

	bool operator>(const Computer& c) const noexcept
	{
		for (int i = 0; i < 4; i++)
		{
			if (registers[i]>c.registers[i])
				return true;
			else if (registers[i]<c.registers[i])
				return false;
		}
		return true;
	}
};

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<Instruction> program;
	Instruction instr;
	while (input >> instr)
	{
		program.emplace_back(move(instr));
	}
	input.close();

	Computer c;
	int64_t i;
	for (i = 0; !c.execute(program, i); i++)
	{
	}
	fmt::print("Part1: {}\n", i);
	return 0;
}

File: day25/day25.py
#!/usr/bin/env python3

import sys

class Instruction(object):
    def __init__(self, op, arg1, arg2):
        self.op = op
        self.arg1 = arg1
        self.arg2 = arg2

    def __repr__(self):
        return "{} {} {}".format(self.op, self.arg1, self.arg2)

class Computer(object):
    def __init__(self):
        self.registers = [0] * 4
        self.ip = 0

    def get(self, addr):
        index = "abcd".find(addr)
        if index < 0:
            return int(addr)
        return self.registers[index]

    def set(self, addr, value):
        index = "abcd".find(addr)
        assert(index >= 0)
        self.registers[index] = value

    def load(self, txt):
        self.program = []
        for line in txt.splitlines():
            lst = line.split(" ")
            if len(lst) < 3:
                lst.append("")
            self.program.append(Instruction(*lst))

    def execute(self, init):
        self.ip = 0
        self.registers = [init, 0, 0, 0]
        last_output = None
        stack = []
        while 0 <= self.ip < len(self.program):
            inst = self.program[self.ip]
            if inst.op == "cpy":
                self.set(inst.arg2, self.get(inst.arg1))
                self.ip += 1
            elif inst.op == "inc":
                self.set(inst.arg1, self.get(inst.arg1)+1)
                self.ip += 1
            elif inst.op == "dec":
                self.set(inst.arg1, self.get(inst.arg1)-1)
                self.ip += 1
            elif inst.op == "jnz":
                if self.get(inst.arg1):
                    self.ip += self.get(inst.arg2)
                else:
                    self.ip += 1
            elif inst.op == "tgl":
                addr = self.ip + self.get(inst.arg1)
                if 0 <= addr < len(self.program):
                    other = self.program[self.ip + self.get(inst.arg1)]
                    if other.op == "inc":
                        other.op = "dec"
                    elif other.op == "dec" or other.op == "tgl":
                        other.op = "inc"
                    elif other.op == "jnz":
                        other.op = "cpy"
                    else:
                        other.op = "jnz"
                self.ip += 1
            elif inst.op == "out":
                value = self.get(inst.arg1)
                if value == last_output:
                    return False
                last_output = value

                # cycle detection
                cur = tuple(self.registers)
                while stack:
                    if stack[-1] == cur:
                        # NOTE: cycle detected
                        return True
                    if stack[-1] > cur:
                        stack.pop()
                    else:
                        break

                stack.append(cur)
                self.ip += 1
            else:
                break

        return self.get("a")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    c = Computer()
    c.load(txt)
    i = 0
    while not c.execute(i):
        i += 1

    print("Part1:", i)
